plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobContainerRetriesTests.java

```diff
@@ -0,0 +1,429 @@
+
+            if ("POST".equals(exchange.getRequestMethod())) {
+                assertThat(params.get("name"), equalTo("write_large_blob"));
+                if (countInits.decrementAndGet() <= 0) {
+                    byte[] response = Streams.readFully(exchange.getRequestBody()).utf8ToString().getBytes(UTF_8);
+                    exchange.getResponseHeaders().add("Content-Type", "application/json");
+                    exchange.getResponseHeaders().add("Location", httpServerUrl() +
+                        "/upload/storage/v1/b/bucket/o?uploadType=resumable&upload_id=" + sessionUploadId.get());
+                    exchange.sendResponseHeaders(RestStatus.OK.getStatus(), response.length);
+                    exchange.getResponseBody().write(response);
+                    exchange.close();
+                    return;
+                }
+
+            } else if ("PUT".equals(exchange.getRequestMethod())) {
+                final String uploadId = params.get("upload_id");
+                if (uploadId.equals(sessionUploadId.get()) == false) {
+                    logger.debug("session id [{}] is gone", uploadId);
+                    assertThat(goneChunk, greaterThan(0));
+                    Streams.readFully(exchange.getRequestBody());
+                    exchange.sendResponseHeaders(HttpStatus.SC_GONE, -1);
+                    exchange.close();
+                    return;
+                }
+
+                if (countUploads.get() == (goneChunk * nbErrors)) {
+                    if (allow410Gone.compareAndSet(true, false)) {
+                        final String newUploadId = UUIDs.randomBase64UUID();
+                        logger.debug("chunk [{}] gone, updating session ids [{} -> {}]", goneChunk, sessionUploadId.get(), newUploadId);
+                        sessionUploadId.set(newUploadId);
+
+                        // we must reset the counters because the whole object upload will be retried
+                        countInits.set(nbErrors);
+                        countUploads.set(nbErrors * totalChunks);
+
+                        Streams.readFully(exchange.getRequestBody());
+                        exchange.sendResponseHeaders(HttpStatus.SC_GONE, -1);
+                        exchange.close();
+                        return;
+                    }
+                }
+
+                final String range = exchange.getRequestHeaders().getFirst("Content-Range");
+                assertTrue(Strings.hasLength(range));
+
+                if (countUploads.decrementAndGet() % 2 == 0) {
+                    final ByteArrayOutputStream requestBody = new ByteArrayOutputStream();
+                    final long bytesRead = Streams.copy(exchange.getRequestBody(), requestBody);
+                    assertThat(Math.toIntExact(bytesRead), anyOf(equalTo(defaultChunkSize), equalTo(lastChunkSize)));
+
+                    final Integer rangeStart = TestUtils.getContentRangeStart(range);
```

Comment:
The range params can just be `int` can't they? (they'll NPE below anyway if there's a `null` return here)

Reply:
Right, thanks